File: 1411-number-of-ways-to-paint-n-3-grid/1411-number-of-ways-to-paint-n-3-grid.cpp
class Solution {
public:
    int mod=1e9+7;
    int dp[50001][4][4][4];
    int memo(int idx,int prevCol1,int prevCol2,int prevCol3){
        if(idx<0)return 1;
        
        if(dp[idx][prevCol1][prevCol2][prevCol3]!=-1)return dp[idx][prevCol1][prevCol2][prevCol3];
        
        int ans=0;
        
        vector<int>colors={1,2,3};
        
        for(int A:colors){
            if(prevCol1==A)continue;
            
            for(int B:colors){
                if(prevCol2==B || A==B)continue;
                
                for(int C:colors){
                    if(prevCol3==C || B==C)continue;
                    
                    ans=(ans+memo(idx-1,A,B,C))%mod;
                    // ans%=mod;
                }
                
            }
        }
        
       return dp[idx][prevCol1][prevCol2][prevCol3]=ans;
        
    }
    
    int numOfWays(int n) {
        memset(dp,-1,sizeof(dp));
        return memo(n-1,0,0,0);
    }
};

File: 1411-number-of-ways-to-paint-n-3-grid/1411-number-of-ways-to-paint-n-3-grid.java
class Solution {
    
    int [][][][]dp;
    int mod=1_000_000_007;
    public int memo(int n,int A,int B,int C){
        if(n<0)return 1;
        if(dp[n][A][B][C]!=-1)return dp[n][A][B][C];
        
        int ans=0;
        
        int []arr= {1,2,3};
        
        for(int a:arr){
            if(a==A)continue;
            for(int b:arr){
                if(b==B || b==a)continue;
                for(int c:arr){
                    if(c==C || c==b)continue;
                    ans=(ans+memo(n-1,a,b,c))%mod;
                }
            }
        }
        return dp[n][A][B][C]=ans;
    }
    
    public int numOfWays(int n) {
        dp=new int[n][4][4][4];
        for(int [][][]i:dp)
            for(int [][]j:i)
                for(int []k:j)
                    Arrays.fill(k,-1);
        return memo(n-1,0,0,0);
    }
}